src/common/snippets/src/lowered/pass/extract_loop_invariants.cpp

```diff
@@ -0,0 +1,244 @@
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "snippets/lowered/pass/extract_loop_invariants.hpp"
+
+#include "snippets/itt.hpp"
+#include "snippets/lowered/linear_ir.hpp"
+#include "snippets/lowered/loop_manager.hpp"
+#include "snippets/snippets_isa.hpp"
+#include "snippets/utils.hpp"
+
+namespace ov {
+namespace snippets {
+namespace lowered {
+namespace pass {
+
+bool ExtractLoopInvariants::run(LinearIR& linear_ir) {
+    OV_ITT_SCOPED_TASK(ov::pass::itt::domains::SnippetsTransform, "Snippets::ExtractLoopInvariants")
+    bool modified = false;
+    const auto& loop_manager = linear_ir.get_loop_manager();
+
+    auto remove_last_loop_id = [](const std::shared_ptr<Expression>& expr) {
+        auto loop_ids = expr->get_loop_ids();
+        loop_ids.pop_back();
+        expr->set_loop_ids(loop_ids);
+    };
+
+    auto get_stride_after_move_outer = [](const LoopPort& loop_port) {
+        const auto& expr_port = loop_port.expr_port;
+        const auto& layout = expr_port->get_descriptor_ptr()->get_layout();
+        const auto& shape = expr_port->get_descriptor_ptr()->get_shape();
+
+        size_t shape_dim_idx = 0;
+        if (expr_port->get_type() == ExpressionPort::Input) {
+            shape_dim_idx = utils::get_input_dim_idx(layout, loop_port.dim_idx);
+        } else if (expr_port->get_type() == ExpressionPort::Output) {
+            shape_dim_idx = utils::get_output_dim_idx(layout, loop_port.dim_idx);
+        } else {
+            OPENVINO_THROW("Unsupported expression port type!");
+        }
+
+        int64_t stride = 1;
+        for (size_t i = shape_dim_idx; i < shape.size(); ++i) {
+            if (utils::is_dynamic_value(shape[i])) {
+                return utils::get_dynamic_value<int64_t>();
+            }
+            stride *= static_cast<int64_t>(shape[i]);
+        }
+        return stride;
+    };
+
+    auto extract_from_loop = [&](const size_t& inner_loop_id) {
```

Comment:
It can be also the separate function 😃 
```
bool extract_from_loop = [&](const size_t& inner_loop_id, LinearIR& linear_ir) {
    ....
    return status; // return True if extraction has been completed
}
```

Then variable `modified` can be updated so (at the moment it's always false)
```
for (size_t i = 0; i < identified_loop_ids.size(); i++) {
         modified |= extract_from_loop(identified_loop_ids[i]);
}
```

Reply:
great! done in [c575707](https://github.com/openvinotoolkit/openvino/pull/24044/commits/c5757078d0c1dbf4e73f824cb21c35a10b6d7cfd). Thanks Alexandra.